### ISSUE TITLE
[Bug]: heilbronn _de isn’t working anymore

### ISSUE THREAD

**Issue by hubelbubel** [mampfes/hacs_waste_collection_schedule#1728]
### I Have A Problem With:

A specific source

### What's Your Problem

It looks like Heilbronn is switching they’re service. The current heilbronn_de isn’t working anymore.

The new service is published under https://abfallwirtschaft.heilbronn.de/

### Checklist Source Error

- [ ] Use the example parameters for your source (often available in the documentation) (don't forget to restart Home Assistant after changing the configuration)
- [ ] Checked that the website of your service provider is still working
- [ ] Tested my attributes on the service provider website (if possible)
- [X] I have tested with the latest version of the integration (master) (for HACS in the 3 dot menu of the integration click on "Redownload" and choose master as version)

### Checklist Sensor Error

- [ ] Checked in the Home Assistant Calendar tab if the event names match the types names (if types argument is used)

### Required

- [X] I have searched past (closed AND opened) issues to see if this bug has already been reported, and it hasn't been.
- [X] I understand that people give their precious time for free, and thus I've done my very best to make this problem as easy as possible to investigate.

**Comment by 5ila5** [comment]
The old Heilbronn source only gets dates till 2024-01-31. 

I tried to take a look at their new calendar, but I could not figure out how to get the dates:

As far as I could figure out they just send requests to https://api.heilbronn.de/garbage-calendar?method=get&datatype=districts which does contain all the addresses and its available waste types, but I could not find any request getting any actual dates, so I guess it's probably part of the JavaScript, but the JavaScript is nearly unreadable as it is "compiled" from dart/flutter. The App does exactly the same: 
![grafik](https://github.com/mampfes/hacs_waste_collection_schedule/assets/38183212/bdb3a3c2-a048-4df5-8498-ff8ac81be231)

So I currently do not see a way to get this source up and running again. You could ask them if there is a way to get an ICS/ICAL file to use with the [ICS source](https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/source/ics.md), but they currently do not support it on their website.

You may be able to use the app to import the events in your calendar (download dropdown only available in the mobile app) and then export them to an ICS file, upload to your HA instance and use the ICS source but this far from ideal.

Or you can use the [User Specified/static source](https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/source/static.md) and define your dates "by hand".

**Comment by ReneNulschDE** [comment]
The dates are here https://api.heilbronn.de/garbage-calendar?method=get&datatype=pickupdates

Workflow would be
1. Load https://api.heilbronn.de/garbage-calendar?method=get&datatype=districts
2. Find address (sometime are * is used for the complete street)
3. Take the collection-calendar id by bin type
4. Load https://api.heilbronn.de/garbage-calendar?method=get&datatype=pickupdates
5. Select the calender based on the data from step 3

**Comment by hubelbubel** [comment]
Thanks @ReneNulschDE and @5ila5 for all you effort.
I tried to build a new script for the new API. When i run the code on my local machine i get a nice collection with data. But it wont work if i copy over the script to home assistant to use it there. Maybe one of you can do a quick check of my code?

```python
from datetime import datetime

import requests
from waste_collection_schedule import Collection  # type: ignore[attr-defined]
from waste_collection_schedule.service.ICS import ICS

TITLE = "Heilbronn Entsorgungsbetriebe"
DESCRIPTION = "Source for city of Heilbronn, Germany."
URL = "https://abfallwirtschaft.heilbronn.de/"
TEST_CASES = {
    "Rosenau": {
        "plz": 74072,
        "strasse": "Rosenau",
    },
    "Biberach": {
        "strasse": "Kehrhüttenstraße",
        "plz": 74078,
    },
    "Klingenberg:": {
        "strasse": "Wittumhalde",
        "plz": "74081",
    },
    "Klingenberg (hausnr as string):": {
        "strasse": "Wittumhalde",
        "plz": "74081",
    },
    "Rosenbergstraße 53": {
        "strasse": "Rosenbergstraße",
        "plz": "74074",
    },
    "Rosenbergstraße 41": {
        "strasse": "Rosenbergstraße",
        "plz": "74072",
    },
}

ICON_MAP = {
    "r": "mdi:trash-can",
    "b": "mdi:leaf",
    "c": "mdi:leaf",
    "l": "mdi:recycle",
    "a": "mdi:package-variant",
    "p": "mdi:package-variant",
}



class Source:
    def __init__(self, plz: int, strasse: str):
        self._plz: str = str(plz)
        self._strasse: str = strasse
        self._ics = ICS()

    def fetch(self):
        r = requests.get(
            "https://api.heilbronn.de/garbage-calendar?method=get&datatype=districts"
        )
        r.raise_for_status()
        data = r.json()

        districts = data["data"][self._plz][self._strasse]["*"]

        # filter waste type
        filteredData = {
            "light-packaging": districts.get("yellow"),
            "paper": districts.get("paper"),
            "bio_2": districts.get("bio"),
            "bio_1": districts.get("bio"),
            "christmastree": districts.get("green"),
            "residual_2": districts.get("residual")
        }

        r = requests.get(
                    "https://api.heilbronn.de/garbage-calendar?method=get&datatype=pickupdates"
                )
        r.raise_for_status()
        pickupDates = r.json()

        entries = []

        for keyDistrict, valueDistrict in filteredData.items():
            value = pickupDates["data"][valueDistrict][keyDistrict]
            for key2, value2 in value.items():
                date: dict = datetime.strptime(value2, "%Y-%m-%d").date()
                entry = keyDistrict
                if entry in ICON_MAP:
                    icon = ICON_MAP.get(entry)
                else:
                    icon = ICON_MAP.get(entry[0])
                entries.append(Collection(date, entry, icon))

                return entries

```

**Comment by ReneNulschDE** [comment]
The indentation of the line "return entries" is wrong. In the current structure you return after the first entry. Bring it to the same level like "for keyD...."

There are multiple things that you need to check:
1. A house number is needed for some streets
2. There more bin types
3. You can filter out old entries. The dateset has entries from the past.

**Comment by 5ila5** [comment]
3. point @ReneNulschDE is not valid 

> - A source script should return all data for all available waste types.
> - A source script should not provide options to limit the returned waste types.
> - A source script should return all data for the entire time period available (including past dates if they are returned).
> - A source script should not provide a configuration option to limit the requested time frame.

https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/contributing_source.md#python-source-script



did some quick changes, needs testing: ([how to test](https://github.com/mampfes/hacs_waste_collection_schedule/blob/master/doc/contributing_source.md#test-the-new-source-file))

```python
from datetime import datetime

import requests
from waste_collection_schedule import Collection  # type: ignore[attr-defined]
from waste_collection_schedule.service.ICS import ICS

TITLE = "Heilbronn Entsorgungsbetriebe"
DESCRIPTION = "Source for city of Heilbronn, Germany."
URL = "https://heilbronn.de"
TEST_CASES = {
    "Klingenberg (hausnr as string):": {
        "strasse": "Wittumhalde",
        "plz": "74081",
        # "hausnr": "75",
    },
    "Rosenbergstraße 53": {
        "strasse": "Rosenbergstraße",
        "plz": "74074",
        "hausnr": "53",
    },
    "Rosenbergstraße 41": {
        "strasse": "Rosenbergstraße",
        "plz": "74072",
        "hausnr": "41",
    },
}

ICON_MAP = {
    "R": "mdi:trash-can",
    "B": "mdi:leaf",
    "GRÜN": "mdi:leaf",
    "G": "mdi:recycle",
    "A": "mdi:package-variant",
    "BT": "mdi:package-variant",
}


class Source:
    def __init__(self, plz: int, strasse: str, hausnr: str | int| None = None):
        self._plz: str = str(plz)
        self._strasse: str = strasse
        self._hausnr: str = str(hausnr) if hausnr is not None else None

    def fetch(self):
        r = requests.get(
            "https://api.heilbronn.de/garbage-calendar?method=get&datatype=districts"
        )
        r.raise_for_status()
        data = r.json()

        street = data["data"][self._plz][self._strasse]
        if not self._hausnr or "*" in street:
            districts:dict = street["*"]
        else:
            districts:dict = street[self._hausnr]

        # filter waste type
        collection_keys = [value for key,value in districts.items() if key not in ("city", "district")]

        r = requests.get(
                    "https://api.heilbronn.de/garbage-calendar?method=get&datatype=pickupdates"
                )
        r.raise_for_status()
        pickupDates = r.json()

        entries = []

        for valueDistrict in collection_keys:
            value = pickupDates["data"][valueDistrict]
            for collection_type, collection_dates in value.items():
                for value2 in collection_dates.values():
                    date: dict = datetime.strptime(value2, "%Y-%m-%d").date()
                    entry = collection_type
                    if entry in ICON_MAP:
                        icon = ICON_MAP.get(entry)
                    else:
                        icon = ICON_MAP.get(entry[0])
                    entries.append(Collection(date, entry, icon))

        return entries
```

**Comment by hubelbubel** [comment]
Thank you for your adjustments. It works nice, but the mapping for the icons are not right anymore. The trash type is now in english. Maybe you can take my mapping:

```python
ICON_MAP = {
    "r": "mdi:trash-can",
    "b": "mdi:leaf",
    "c": "mdi:leaf",
    "g": "mdi:leaf",
    "l": "mdi:recycle",
    "a": "mdi:package-variant",
    "p": "mdi:package-variant",
}
```

**Comment by Eisbaer2** [comment]
Was muss ich tun, um jetzt wieder Benachrichtigungen via Home Assistant erhalten zu können? In der Doku hat sich noch nichts geändert.

**Comment by derDeno** [comment]
Is this new source already merged or is there already a pr?

**Comment by 5ila5** [comment]
Fixed by #1815 

This will be part of release 1.46.0 or is available now as master (git) version (3-dot menu -> redownload -> change version to master -> fully restart HA (this will increase update cycle by a lot)

You should not need to change the configuration. But they seemed to remove some house numbers? One of our Test cases did not work anymore and needed a different house number.

The collection type names changed as well

**Comment by 5ila5** [comment]
@Eisbaer2 1.46.0 is not released yet. Changing version to "master" (current state of this reposetory) should fix the issue. If it does not there is probably something wrong with HACS you can update the changed file of the PR manually. And wait for version 1.46.0

**Comment by Eisbaer2** [comment]
@5ila5: works fine again. Only every entry is there 5x identical. In the sensor as well as in the calendar.
Any ideas why?

**Comment by 5ila5** [comment]
Fixed by #1839. You just need to update the master version.
